Add Chinese translation of pacemaker-fenced files

daaitudian opened the pull request:
Hi:
In this pr, I added Chinese translation about pacemaker-fenced files.

wenningerk commented:
test this please

kgaillot commented:
I just realized the way we started doing option translations is incorrect. The translation of tool error output is fine. But the option meta-data is different because the meta-data XML format inherently supports translations with the lang attribute. Currently, for translated option meta-data, the output changes from

    <longdesc lang="en"> ... English translation ... </longdesc>

to

    <longdesc lang="en"> ... Chinese translation ... </longdesc>

That means the English translation is lost, and the Chinese translation has an incorrect lang attribute. The intended use of the meta-data XML with translations is multiple elements like:

    <longdesc lang="en"> ... English translation ... </longdesc>
    <longdesc lang="zh"> ... Chinese translation ... </longdesc>
    ... etc. ...

This is a bit of a challenge since the XML is generated by the C code. In the case of the fencer options translated in this PR, a quick fix would be to modify the C code like:

         printf(" <version>1.0</version>\n");
         printf(" <longdesc lang=\"en\">Instance attributes available for all \"stonith\"-class resources"
                                        " and used by Pacemaker's fence daemon, formerly known as stonithd</longdesc>\n");
         printf(" <shortdesc lang=\"en\">Instance attributes available for all \"stonith\"-class resources</shortdesc>\n");
+#ifdef ENABLE_NLS
+       printf(_(" <longdesc lang=\"en\">Instance attributes available for all \"stonith\"-class resources"
+                                      " and used by Pacemaker's fence daemon, formerly known as stonithd</longdesc>\n"));
+       printf(_(" <shortdesc lang=\"en\">Instance attributes available for all \"stonith\"-class resources</shortdesc>\n"));
+#endif
         printf(" <parameters>\n");

That way, both the English version and the translated version will be shown. Unfortunately, if there is no translation available, the English version would be shown twice, but I don't see an easy way around that. Also, it wouldn't show all available translations, just whatever was set in the user's locale. But I think those are acceptable limitations for now.

A similar approach could be taken for the existing translations formatted by pcmk__format_option_metadata(). In that case we would need to keep the English translation without gettext markers, and if NLS is enabled, add an additional g_string_append_printf() for the English translation with gettext markers.

That is my best suggestion for a quick fix. As we get more experience with this, we may be able to come up with something better.

daaitudian commented:
Hi @kgaillot : 
 
Thank you for your reply.
Regarding your question, I don't understand something. I did some tests. When using the enable-nls option, we also modified the corresponding XML tag, and replaced the value of lang from en to zh in the translation file.

po/zh_CN.po :
#: daemons/fenced/pacemaker-fenced.c:
#, c-format
msgid ""
" <longdesc lang=\"en\">... English translation ...</longdesc>\n"
msgid ""
" <longdesc lang=\"zh\">... Chinese translation ... </longdesc>\n"

So when my LANG is zh_CN.UTF-8, the display content is:

<longdesc lang="zh">... Chinese translation ...</longdesc>

When I change LANG to en_US.UTF-8, the display changes back to:

<longdesc lang="en">...English translation...</longdesc>

or

Do you want to keep one more English translation of the same sentence, for example:

export LANG=zh_CN.UTF-8
<longdesc lang="en">...English translation...</longdesc>
<longdesc lang="zh">... Chinese translation for ... </longdesc>

Looking forward to your reply.

gao-yan commented on po/zh_CN.po:
```diff
@@ -49,12 +49,71 @@ msgstr ""
+"additional parameters.\n"
+"     </longdesc>\n"
+msgstr ""
+"    <longdesc lang=\"zh\">某些设备可能不支持使用标准的端口参数，也可能会提供额外的端口参数。\n"
```
`'port'` would probably better remain untranslated, since it's the literal name of the parameter.

gao-yan commented on po/zh_CN.po:
```diff
@@ -49,12 +49,71 @@ msgstr ""
+"    <longdesc lang=\"en\">Eg. node1:1;node2:2,3 would tell the cluster to "
+"use port 1 for node1 and ports 2 and 3 for node2</longdesc>\n"
+msgstr ""
+"    <longdesc lang=\"zh\">例如：pcmk_host_map=\"node:1;node2:2,3\"表示让集群的"
```
Probably would be more precise with "...表示让集群为节点..."?

gao-yan commented on po/zh_CN.po:
```diff
@@ -49,12 +49,71 @@ msgstr ""
+"    <shortdesc lang=\"en\">Advanced use only: An alternate parameter to "
+"supply instead of 'port'</shortdesc>\n"
+msgstr ""
+"    <shortdesc lang=\"zh\">Advanced use only:(仅限高级使用)备用参数可替代'port'</shortdesc>\n"
```
Is it intentional to repeat "Advanced use only" here given that it's already translated?

kgaillot commented:
> Hi @kgaillot :
> 
> Thank you for your reply. Regarding your question, I don't understand something. I did some tests. When using the enable-nls option, we also modified the corresponding XML tag, and replaced the value of lang from en to zh in the translation file.

Right, for the translations in this PR. But for the original translations that were merged, using pcmk__format_option_metadata(), the "lang" attribute is not changed, because of how that function works.

> 
> ```
> po/zh_CN.po :
> #: daemons/fenced/pacemaker-fenced.c:
> #, c-format
> msgid ""
> " <longdesc lang=\"en\">... English translation ...</longdesc>\n"
> msgid ""
> " <longdesc lang=\"zh\">... Chinese translation ... </longdesc>\n"
> ```
> 
> So when my LANG is zh_CN.UTF-8, the display content is:
> 
> `<longdesc lang="zh">... Chinese translation ...</longdesc>`
> 
> When I change LANG to en_US.UTF-8, the display changes back to:
> 
> `<longdesc lang="en">...English translation...</longdesc>`
> 
> or
> 
> Do you want to keep one more English translation of the same sentence, for example:
> 
> export LANG=zh_CN.UTF-8
> 
> ```
> <longdesc lang="en">...English translation for Instance attributes available...</longdesc>
> <longdesc lang="zh">...Chinese translation for Instance attributes available...</longdesc>
> ```

Correct. The XML meta-data follows the OCF resource agent schema, which provides for multiple translations in that manner. I think that providing the English translation and the translation in the user's locale is sufficient. (In other words, if we one day have multiple translations, I don't think we need to show every translation, just those two.)

> 
> Looking forward to your reply.

daaitudian commented:
Thanks @kgaillot , I get it. I would like to know, if we mark the 'en' in PCMK__FORMAT_OPTION_METADATA, then we change it into zh in zh.Po file, is this feasible?

E.g:
```
Char *
PCMK__FORMAT_OPTION_METADATA (Const Char * Name, Const Char * Desc_short,
                              Const char * desc_long,
                              PCMK__cluster_option_t * Option_List, Int Len
{
... etc ...
     g_string_append_printf (s, "<? XML Version = \" 1.0 \ "?>"
                               "<! Doctype resource-agent system \" ra-API-1.DTD \ "> \ n"
                               <Resource-Agent Name = \ "% s \"> \ n "
                               "<version>% s </ version> \ n"
                               _ ("<longdesc lang = \" en \ ">% s </ longdesc> \ n")
                               _ ("<shortdesc lang = \" en \ ">% s </ shortdesc> \ n")
                               "<parameters> \ n",
                               Name, PCMK_OCF_VERSION, ESCAPED_LONG, ESCAPED_SHORT;
```
po/zh_CN.po :
```
#: lib/common/options.c:574
#, c-format
msgid "  <longdesc lang=\"en\">%s</longdesc>\n"
msgstr "  <longdesc lang=\"zh\">%s</longdesc>\n"

#: lib/common/options.c:575
#, c-format
msgid "  <shortdesc lang=\"en\">%s</shortdesc>\n"
msgstr "  <shortdesc lang=\"zh\">%s</shortdesc>\n"

```

kgaillot commented:
> Thanks @kgaillot , I get it. I would like to know, if we mark the 'en' in PCMK__FORMAT_OPTION_METADATA, then we change it into zh in zh_CN.po file, is this feasible?
> 
> E.g:
> 
> ```
> Char *
> PCMK__FORMAT_OPTION_METADATA (Const Char * Name, Const Char * Desc_short,
>                               Const char * desc_long,
>                               PCMK__cluster_option_t * Option_List, Int Len
> {
>      ... etc ...
>      g_string_append_printf (s, "<? XML Version = \" 1.0 \ "?>"
>                                "<! Doctype resource-agent system \" ra-API-1.DTD \ "> \ n"
>                                <Resource-Agent Name = \ "% s \"> \ n "
>                                "<version>% s </ version> \ n"
>                                _("<longdesc lang = \" en \ ">% s </ longdesc> \ n")
>                                _("<shortdesc lang = \" en \ ">% s </ shortdesc> \ n")

The above would translate every longdesc/shortdesc lang attribute, even for options that haven't been translated yet. Also, the "en" translation would not be shown -- for meta-data it would be best to show both "en" and the user's locale since it supports that.

For this function, I was thinking along the lines of

            g_string_append_printf(s, "    <parameter name=\"%s\">\n",
                                      option_list[lpc].name);

            // First show the "en" version without translation
            escaped_long = crm_xml_escape(option_list[lpc].description_long?
                                             option_list[lpc].description_long :
                                              option_list[lpc].description_short);
            escaped_short = crm_xml_escape(option_list[lpc].description_short);
    
            g_string_append_printf(s,
                                      "      <longdesc lang=\"en\">%s%s%s</longdesc>\n"
                                      "      <shortdesc lang=\"en\">%s</shortdesc>\n",
                                      escaped_long,
                                      (option_list[lpc].values? "  Allowed values: " : ""),
                                      (option_list[lpc].values? option_list[lpc].values : ""),
                                      escaped_short);

    #ifdef ENABLE_NLS
            // Now show the translation for the user's locale
            escaped_long = crm_xml_escape(option_list[lpc].description_long?
                                             _(option_list[lpc].description_long) :
                                              _(option_list[lpc].description_short));
            escaped_short = crm_xml_escape(_(option_list[lpc].description_short));
    
            locale = setlocale(NULL); // Get current locale as const char *
            g_string_append_printf(s,
                                      "      <longdesc lang=\"%s\">%s%s%s</longdesc>\n"
                                      "      <shortdesc lang=\"%s\">%s</shortdesc>\n",
                                      locale, escaped_long,
                                      (option_list[lpc].values? "  Allowed values: " : ""),
                                      (option_list[lpc].values? option_list[lpc].values : ""),
                                      locale, escaped_short);
    #endif

With the above, the localized version would be shown even if there's no translation, but we could get around that by keeping the first escaped\_short/escaped\_long and only adding the translation if the second one is different (strcmp()).

daaitudian commented:
Ok, I will revise and submit again.

daaitudian commented:
Hi  @kgaillot :
Thank you for your help, I have completed the modification. 
I would like to know if there is a way to invalidate ENABLE_NLS when the local user language po file is missing. So that we can output native translations that avoid duplicates.

kgaillot commented on lib/common/options.c:
```diff
@@ -584,20 +585,40 @@ pcmk__format_option_metadata(const char *name, const char *desc_short,
                                   (option_list[lpc].values? "  Allowed values: " : ""),
                                   (option_list[lpc].values? option_list[lpc].values : ""),
                                   escaped_short);
-        free(escaped_long);
```
escaped_long and escaped_short do still need to be freed before reusing them for NLS (though it's probably better to use different variables for NLS, so we can compare them to avoid duplicates)

kgaillot commented on lib/common/options.c:
```diff
@@ -584,20 +585,40 @@ pcmk__format_option_metadata(const char *name, const char *desc_short,
+                                  locale,
+				  escaped_short);
+#endif
+	free(escaped_long);
```
The second free of escaped_long/escaped_short (or the free of new variables) would be inside the #endif

daaitudian commented:
Thanks. I've changed the location of the variable free.

kgaillot commented:
@daaitudian , are you ready for this to be merged, or did you want to address more of the comments on the translated strings?

daaitudian commented:
I want to merge these first.

kgaillot commented:
OK, thanks!
